File: examples2d/rope_joints2.rs
use crate::utils::character;
use crate::utils::character::CharacterControlMode;
use rapier_testbed2d::Testbed;
use rapier2d::control::{KinematicCharacterController, PidController};
use rapier2d::prelude::*;

pub fn init_world(testbed: &mut Testbed) {
    /*
     * World
     */
    let mut world = PhysicsWorld::new();

    /*
     * Ground
     */
    let ground_size = 0.75;
    let ground_height = 0.1;

    let rigid_body = RigidBodyBuilder::fixed().translation(Vector::new(0.0, -ground_height));
    let collider = ColliderBuilder::cuboid(ground_size, ground_height);
    let _ = world.insert(rigid_body, collider);

    let rigid_body = RigidBodyBuilder::fixed()
        .translation(Vector::new(-ground_size - ground_height, ground_size));
    let collider = ColliderBuilder::cuboid(ground_height, ground_size);
    let _ = world.insert(rigid_body, collider);

    let rigid_body = RigidBodyBuilder::fixed()
        .translation(Vector::new(ground_size + ground_height, ground_size));
    let collider = ColliderBuilder::cuboid(ground_height, ground_size);
    let _ = world.insert(rigid_body, collider);

    /*
     * Character we will control manually.
     */
    let rigid_body =
        RigidBodyBuilder::kinematic_position_based().translation(Vector::new(0.0, 0.3));
    let collider = ColliderBuilder::cuboid(0.15, 0.3);
    let (character_handle, _) = world.insert(rigid_body, collider);

    /*
     * Tethered Ball
     */
    let rad = 0.04;

    let rigid_body =
        RigidBodyBuilder::new(RigidBodyType::Dynamic).translation(Vector::new(1.0, 1.0));
    let collider = ColliderBuilder::ball(rad);
    let (child_handle, _) = world.insert(rigid_body, collider);

    let joint = RopeJointBuilder::new(2.0).local_anchor2(Vector::new(0.0, 0.0));
    world.insert_impulse_joint(character_handle, child_handle, joint);

    /*
     * Callback to update the character based on user inputs.
     */
    let mut control_mode = CharacterControlMode::Kinematic(0.1);
    let mut controller = KinematicCharacterController::default();
    let mut pid = PidController::default();

    testbed.add_callback(move |graphics, physics, _, _| {
        if let Some(graphics) = graphics {
            character::update_character(
                graphics,
                physics,
                &mut control_mode,
                &mut controller,
                &mut pid,
                character_handle,
            );
        }
    });

    /*
     * Set up the testbed.
     */
    testbed.set_physics_world(world);
    testbed.look_at(Vec2::new(0.0, 1.0), 100.0);
}

File: python/rapier-testbed/rapier_testbed/examples2/rope_joints2.py
"""Port of examples2d/rope_joints2.rs.

The original example wires a manually-driven character via
``utils::character``. The Python port keeps the static scene (tether on a
kinematic character body) but omits the keyboard-driven controller; this
keeps the example useful for debug-rendering and headless smoke testing.
"""
from __future__ import annotations

import rapier2d as rp
from .._registry import register

CATEGORY = "Joints"
NAME = "Rope Joints"


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    ground_size = 0.75
    ground_height = 0.1

    handle = bodies.insert(rp.RigidBody.fixed().translation((0.0, -ground_height)))
    colliders.insert_with_parent(
        rp.Collider.cuboid(ground_size, ground_height), handle, bodies
    )

    handle = bodies.insert(
        rp.RigidBody.fixed().translation((-ground_size - ground_height, ground_size))
    )
    colliders.insert_with_parent(
        rp.Collider.cuboid(ground_height, ground_size), handle, bodies
    )

    handle = bodies.insert(
        rp.RigidBody.fixed().translation((ground_size + ground_height, ground_size))
    )
    colliders.insert_with_parent(
        rp.Collider.cuboid(ground_height, ground_size), handle, bodies
    )

    character_handle = bodies.insert(
        rp.RigidBody.kinematic_position_based().translation((0.0, 0.3))
    )
    colliders.insert_with_parent(
        rp.Collider.cuboid(0.15, 0.3), character_handle, bodies
    )

    rad = 0.04
    child_handle = bodies.insert(rp.RigidBody.dynamic().translation((1.0, 1.0)))
    colliders.insert_with_parent(rp.Collider.ball(rad), child_handle, bodies)

    joint = rp.RopeJoint.builder(2.0).local_anchor2((0.0, 0.0))
    impulse_joints.insert(character_handle, child_handle, joint, wake_up=True)

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.set_camera_2d(center=(0.0, 1.0), zoom=100.0)


register(CATEGORY, NAME, init_world, dim=2)

if __name__ == "__main__":
    from .._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
